haskell/test/SMT/Model/Proposal.hs

```diff
@@ -0,0 +1,416 @@
+  canBeFlushed <- isLevelReached proposal =<< (getConfig <&> cProposalFlushLevel)
+  if isExpired then
+    throwError EXPIRED_PROPOSAL
+  else if canBeFlushed && (counter & cCurrent) < (counter & cTotal) then do
+    let counter_ = counter { cCurrent = (counter & cCurrent) + 1}
+    cond <- getStore <&> \s ->
+              doTotalVoteMeetQuorumThreshold proposal s && ((proposal & plUpvotes) > (proposal & plDownvotes))
+
+    unfreezeProposerAndVoterToken cond proposal
+    case cond of
+      True -> do
+        (_, newExtra) <- getStore >>= \s -> ((config & cDecisionLambda) # pair) -$$? (proposal, (s & sExtra))
+        modifyStore $ \s -> s { sExtra = newExtra }
+      False -> pure ()
+    -- skip operations
+    deleteProposal proposalLvl proposalKey
+
+    pure counter_
+  else
+    pure counter
+
+-- | Counter type used in `flush`
+data Counter = Counter {cCurrent :: Natural, cTotal :: Natural}
+  deriving stock (Show)
+
+applyFlush :: BaseDaoContractCall -> Natural -> ModelT ()
+applyFlush _ param = do
+  -- store <- getStore
+  if param == 0 then
+    throwError BAD_ENTRYPOINT_PARAMETER
+  else do
+    let counter :: Counter = Counter {cCurrent = 0, cTotal = param}
+
+    updatedCounter <- getStore >>= \s -> foldM flushOne counter (s & sProposalKeyListSortByDate)
+
+    if (updatedCounter & cCurrent) == 0 then throwError EMPTY_FLUSH
+    else pure ()
+
+  where
+    flushOne :: Counter -> (Natural, ProposalKey) -> ModelT Counter
+    flushOne counter (proposalLvl, proposalKey) = do
+      handleProposalIsOver proposalLvl proposalKey counter
+
+
+applyDropProposal :: BaseDaoContractCall -> ProposalKey -> ModelT ()
+applyDropProposal cc proposalKey = do
+
+  proposal <- checkIfProposalExist proposalKey
+  proposalIsExpired <- isLevelReached proposal =<< (getConfig <&> cProposalExpiredLevel)
+
+  store <- getStore
+  if   ((cc & ccSender) == (proposal & plProposer))
+    || ((cc & ccSender) == (store & sGuardian) && (cc & ccSender) /= (cc & ccSource))
+    || proposalIsExpired
+  then do
+    unfreezeProposerAndVoterToken False proposal
+    deleteProposal (proposal & plStartLevel) proposalKey
+  else
+    throwError DROP_PROPOSAL_CONDITION_NOT_MET
+
+---------------------------------------------------------------
+-- Delegate
+---------------------------------------------------------------
+
+applyUpdateDelegate :: BaseDaoContractCall -> [DelegateParam] -> ModelT ()
+applyUpdateDelegate cc params = do
+  mapM_ updateDelegate params
+  where
+    updateDelegate :: DelegateParam -> ModelT ()
+    updateDelegate param = do
+      delegates <- getStore <&> (\s -> s & sDelegates & T.unBigMap)
+      let key = Delegate { dOwner = cc & ccSender, dDelegate = param & dpDelegate }
+      let newDelegates = BigMap $ case param & dpEnable of
+            True -> Map.insert key () (delegates)
+            False -> Map.delete key (delegates)
+
+      modifyStore $ \s -> s { sDelegates = newDelegates}
```

Comment:
Here's another example, this is currently basically a straight copy of the LIGO source, but you should instead write it as you'd write original Haskell code, based on the result that is expected.

E.g. I would write this something like:
```suggestion
updateDelegate :: BaseDaoContractCall -> Set Delegate -> DelegateParam -> Set Delegate
updateDelegate cc delegates param =
  if param & dpEnable
    then Set.insert delegate delegates
    else Set.delete delegate delegates
  where
    delegate = Delegate { dOwner = cc & ccSender, dDelegate = param & dpDelegate }

applyUpdateDelegate :: BaseDaoContractCall -> [DelegateParam] -> ModelT ()
applyUpdateDelegate cc params = do
  delegates <- getStore <&> (Map.keysSet . T.unBigMap . sDelegates)
  let newDelegates = foldl' (updateDelegate cc) delegates params
  modifyStore $ \s -> s { sDelegates = BigMap $ Map.fromSet (const ()) newDelegates }
```

Reply:
Updated.